src/node_options.h

```diff
@@ -137,6 +139,10 @@ class PerProcessOptions {
 #endif
 #endif
 
+#if defined(NODE_REPORT)
+  std::string cmdline;
```

Comment:
At this point I would like to seek your views on which structure / class are perfect for some of these data elements: 

we have:
```C++
  bool exception_hook_initialised = false;
  bool error_hook_initialised = false;
  bool signal_thread_initialised = false;
  std::string report_filename;
  std::string report_directory;
  std::string cmdline;
```

and possibly 
```c
static bool report_active = false;
```
also joining the troop. To accommodate those, we do have 
```js
IsolateData,
PerIsolateOptions,
EnvironmentOptions,
PerProcessOptions
```
these placeholders. I am bit unsure what are the optimal placements for inter-thread and inter-isolate safety and isolation.

Reply:
Only for `cmdline`: As we are populating this from src/node.cc at such an early stage when `env` or `isolatedata` is unavailable or inaccessible. So per_process_options seem to be the only way for accessing it.

Reply:
Okay, so…

> `bool exception_hook_initialised = false;`
`bool error_hook_initialised = false;`

These would be on `IsolateData`, if the individual hooks added by node-report stick around, because they track information about the V8 isolate.

> bool signal_thread_initialised = false;

I’m not sure. Would we want the ability to spin up signal handlers from every isolate? From every thread? Signal handlers are more or less global state, which is why they are one of the few APIs disabled in our built-in `Worker`s implementation…

I guess this problem “goes away” if we switch to the JS variant of this? I don’t know where the comment was, but if `process.on('SIGFOO', () => generateReport());` does the trick, we can ignore this.

> `std::string report_filename;`
`std::string report_directory;`

I think their current place (`IsolateData`) is fine. Ideally, I think they should be per-`Environment`, but given that we need to access these even when there is no single specific `Environment` available, this might make sense.

> std::string cmdline;

I was going to suggest a global variable, but yeah, essentially `per_process_options` would be fine.

> bool report_active = false;

(See https://github.com/nodejs/node/pull/22712#discussion_r217527517)

Reply:
@addaleax - done ; with signals and fatalerrors fully controlled from the JS land, `exception_hook_initialised` and `signal_thread_initialised` are now JS fields, and taken away from C++ code. The rest, restructured as proposed.